engine/schema/src/com/cloud/storage/dao/StoragePoolTagsDaoImpl.java

```diff
@@ -77,4 +90,68 @@ public void deleteTags(long poolId) {
         txn.commit();
     }
 
+    @Override
+    public List<StoragePoolTagVO> searchByIds(Long... stIds) {
+        final int detailsBatchSize = getDetailsBatchSize();
+
+        // query details by batches
+        List<StoragePoolTagVO> uvList = new ArrayList<StoragePoolTagVO>();
+        int curr_index = 0;
+
+        if (stIds.length > detailsBatchSize) {
+            while ((curr_index + detailsBatchSize) <= stIds.length) {
+                searchForStoragePoolIdsInternal(curr_index, detailsBatchSize, stIds, uvList);
+                curr_index += detailsBatchSize;
+            }
+        }
+
+        if (curr_index < stIds.length) {
+            int batch_size = (stIds.length - curr_index);
+            searchForStoragePoolIdsInternal(curr_index, batch_size, stIds, uvList);
+        }
+
+        return uvList;
+    }
+
+    /**
+     * Search storage pools
+     * @param currIndex current index
+     * @param batchSize batch size
+     * @param stIds storage tags array
+     * @param uvList
+     */
+    protected void searchForStoragePoolIdsInternal(int currIndex, int batchSize, Long[] stIds, List<StoragePoolTagVO> uvList) {
```

Comment:
Excellent improvement, I copied almost exactly as you wrote it :) I also added some description on `uvList` parameter, which also renamed to `pools`